File: 0112-path-sum/0112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
function solve(sum, targetSum, root) {
    if (root==null){
        return false;
    }
    if (root.left == null && root.right == null && sum + root.val == targetSum) {
        return true;
    }
    return (solve(sum + root.val, targetSum, root.left) || solve(sum + root.val, targetSum, root.right))


}
var hasPathSum = function (root, targetSum) {

    var sum = 0;
    return solve(sum, targetSum, root);

};

File: 0112-path-sum/0112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode* root, int targetSum, int sum, bool& flag) {
        if (root == NULL) {
            return;
        }

        if (root->left == NULL && root->right == NULL) {
            if (sum + root->val == targetSum) {
                flag = true;
            }
            return;
        }
        // left path explore karo
        solve(root->left, targetSum, sum + root->val, flag);
        // right path karo
        solve(root->right, targetSum, sum + root->val, flag);
    }
    bool hasPathSum(TreeNode* root, int targetSum) {
        bool flag = false;
        int sum = 0;
        solve(root, targetSum, sum, flag);

        return flag;
    }
};